File: 007 - 10001st prime/C++/pe_007.cc
/*
 * file:    pe_007.cc
 * title:   10001st prime
 * author:  Alexander P. Gosselin
 * e-mail:  alexandergosselin@gmail.com
 * date:    October 24, 2014
 * 
 * math:    This solution uses the sieve of Sundaram to find the nth
 *          prime p_n. The sieve is initialized using the fact that p_n
 *          is less than n*(ln(n) + ln(ln(n))).
 * 
 *          References:
 *          en.wikipedia.org/wiki/Sieve_of_Sundaram
 *          en.wikipedia.org/wiki/Prime_number_theorem
 *              #Approximations_for_the_nth_prime_number
 */

#include <cmath>
#include <iostream>
#include <vector>
using namespace std;

const int N = 10001; // Nth prime

int main()
{
  int max = floor(N*(log(N) + log(log(N))));
  vector<bool> marked(max/2, false);
  for (size_t i = 1; i < marked.size(); i++) {
    for (size_t j = 1; j <= i; j++) {
      size_t m = i + j + 2*i*j;
      if (m < marked.size()) {
        marked[m] = true;
      } else {
        break;
      }
    }
  }
  int n = 1;
  for (size_t m = 1; m < marked.size(); m++) {
    if (not marked[m]) n++;
    if (n == N) {
      cout << 2*m + 1 << endl;
      break;
    }
  }
  return 0;
}

File: 007 - 10001st prime/Haskell/pe_007.hs
{- |
file:   pe_007.hs
title:  10001st Prime
author: Alexander P. Gosselin
e-mail: alexandergosselin@gmail.com
date:   May 23, 2015

link:   https://projecteuler.net/problem=7

references:
  O'Neill, Melissa E. "The Genuine Sieve of Eratosthenes." 
  J. Funct. Prog. Journal of Functional Programming (2008).
  Web. 23 May 2015. 
  <https://www.cs.hmc.edu/~oneill/papers/Sieve-JFP.pdf>

  "Sieve of Eratoesthenes." Wikipedia. Wikimedia Foundation. Web. 23 May 2015.
  <http://en.wikipedia.org/wiki/Sieve_of_Eratosthenes>

copyright:  (c) 2008 Richard Bird
            Used without permission
            From "The Genuine Sieve of Eratosthenes", page 11
-}

primes = 2:subtract' [3..] composites
    where
        composites = union [multiples p | p <- primes]

multiples n = map (n*) [n..]

subtract' (x:xs) (y:ys) 
    | x < y   = x:subtract' xs (y:ys)
    | x == y  = subtract' xs ys
    | x > y   = subtract' (x:xs) ys

union = foldr merge []
    where
        merge (x:xs) ys = x:merge' xs ys
        merge' (x:xs) (y:ys)
            | x < y   = x:merge' xs (y:ys)
            | x == y  = x:merge' xs ys
            | x > y   = y:merge' (x:xs) ys

pe_007 = last $ take 10001 primes

main :: IO ()
main = do
    print pe_007
